From 68634c5eb09b68782d940e459cd9eaa911df934e Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Fri, 22 Aug 2025 17:37:19 +0800
Subject: [PATCH 215/254] mtk: wpa_supplicant: Add wpa_sm cb functions for mlo
 ft roaming

Add two function for calcualting mlo ft mic.
1. ft_get_valid_links used for get valid link map from reassoc resp.
2. ft_get_link_addr used for get link address by link id.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>

Add support for getting sta link address.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 src/rsn_supp/wpa.h         |  3 +++
 src/rsn_supp/wpa_i.h       | 16 ++++++++++++++++
 wpa_supplicant/wpas_glue.c | 28 ++++++++++++++++++++++++++++
 3 files changed, 47 insertions(+)

diff --git a/src/rsn_supp/wpa.h b/src/rsn_supp/wpa.h
index a33f3679c..b471e94b7 100644
--- a/src/rsn_supp/wpa.h
+++ b/src/rsn_supp/wpa.h
@@ -57,6 +57,9 @@ struct wpa_sm_ctx {
 			     size_t ies_len);
 	int (*send_ft_action)(void *ctx, u8 action, const u8 *target_ap,
 			      const u8 *ies, size_t ies_len);
+	u16 (*ft_get_valid_links)(void *ctx, u8 *assoc_link_id);
+	int (*ft_get_link_addr)(void *ctx, u8 *link_addr, u8 link_id,
+				bool is_assoc_resp);
 	int (*mark_authenticated)(void *ctx, const u8 *target_ap);
 #ifdef CONFIG_TDLS
 	int (*tdls_get_capa)(void *ctx, int *tdls_supported,
diff --git a/src/rsn_supp/wpa_i.h b/src/rsn_supp/wpa_i.h
index 4f75e3cd7..2608feff0 100644
--- a/src/rsn_supp/wpa_i.h
+++ b/src/rsn_supp/wpa_i.h
@@ -373,6 +373,22 @@ static inline int wpa_sm_send_ft_action(struct wpa_sm *sm, u8 action,
 	return -1;
 }
 
+static inline u16 wpa_sm_get_valid_links(struct wpa_sm *sm, u8 *assoc_link_id)
+{
+	if (sm->ctx->ft_get_valid_links)
+		return sm->ctx->ft_get_valid_links(sm->ctx->ctx, assoc_link_id);
+	return 0;
+}
+
+static inline int wpa_sm_get_link_addr(struct wpa_sm *sm, u8 *link_addr, u8 link_id,
+				       bool is_assoc_resp)
+{
+	if(sm->ctx->ft_get_link_addr)
+		return sm->ctx->ft_get_link_addr(sm->ctx->ctx, link_addr, link_id,
+						 is_assoc_resp);
+	return -1;
+}
+
 static inline int wpa_sm_mark_authenticated(struct wpa_sm *sm,
 					    const u8 *target_ap)
 {
diff --git a/wpa_supplicant/wpas_glue.c b/wpa_supplicant/wpas_glue.c
index a1927374c..3ff0c7738 100644
--- a/wpa_supplicant/wpas_glue.c
+++ b/wpa_supplicant/wpas_glue.c
@@ -749,6 +749,32 @@ static int wpa_supplicant_send_ft_action(void *ctx, u8 action,
 }
 
 
+static u16 wpa_supplicant_ft_get_valid_links(void *ctx, u8 *assoc_link_id)
+{
+	struct wpa_supplicant *wpa_s = ctx;
+
+	if (assoc_link_id)
+		*assoc_link_id = wpa_s->mlo_assoc_link_id;
+	return wpa_s->valid_links;
+}
+
+
+static int wpa_supplicant_ft_get_link_addr(void *ctx, u8 *link_addr, u8 link_id,
+					   bool is_assoc_resp)
+{
+	struct wpa_supplicant *wpa_s = ctx;
+
+	if (!link_addr || link_id > MAX_NUM_MLD_LINKS - 1)
+		return -1;
+
+	if (is_assoc_resp)
+		os_memcpy(link_addr, wpa_s->links[link_id].bssid, ETH_ALEN);
+	else
+		os_memcpy(link_addr, wpa_s->links[link_id].addr, ETH_ALEN);
+	return 0;
+}
+
+
 static int wpa_supplicant_mark_authenticated(void *ctx, const u8 *target_ap)
 {
 	struct wpa_supplicant *wpa_s = ctx;
@@ -1496,6 +1522,8 @@ int wpa_supplicant_init_wpa(struct wpa_supplicant *wpa_s)
 #ifdef CONFIG_IEEE80211R
 	ctx->update_ft_ies = wpa_supplicant_update_ft_ies;
 	ctx->send_ft_action = wpa_supplicant_send_ft_action;
+	ctx->ft_get_valid_links = wpa_supplicant_ft_get_valid_links;
+	ctx->ft_get_link_addr = wpa_supplicant_ft_get_link_addr;
 	ctx->mark_authenticated = wpa_supplicant_mark_authenticated;
 #endif /* CONFIG_IEEE80211R */
 #ifdef CONFIG_TDLS
-- 
2.45.2

